src/Indigo/Contracts/HomebaseLite/Impl/Vote.hs

```diff
@@ -1,52 +1,103 @@
 -- SPDX-FileCopyrightText: 2022 Tezos Commons
 -- SPDX-License-Identifier: LicenseRef-MIT-TC
 
-{-# OPTIONS_GHC -Wno-redundant-constraints -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# LANGUAGE RebindableSyntax #-}
 
 module Indigo.Contracts.HomebaseLite.Impl.Vote
   ( propose
   , vote
   , verifyMinBalance
   ) where
 
-import Indigo hiding ((*))
+import Indigo hiding (errorDoc, (*))
 
 import Morley.Util.Interpolate (itu)
 
-propose :: IndigoEntrypoint tp
-propose _ = do
+import Indigo.Contracts.HomebaseLite.Impl.Utils
+import Indigo.Contracts.HomebaseLite.Types
+
+propose :: HasSideEffects => FA2Config -> HomebaseLiteEntrypoint ("proposal_uri" :! URI, "choices" :! [MText])
+propose FA2Config{..} par = do
   description [itu|
     Called by a `participant` to submit a new proposal.
     Takes as input an IPFS URI and a list of strings for the `n`  available choices.
     A new proposal for the given IPFS will be initialized (with a `submitted` status).
     |]
-  pass
+  choices <- new$ par #! #choices
+  when (size choices == 0 nat) do
+    failCustomNoArg @() #emptyChoices
+  uri <- new$ fst par
+  proposals <- new$ storage #! #sProposals
+  when (mem uri proposals) do
+    failCustomNoArg @() #duplicateProposal
+  selfRef <- selfCalling @Parameter (Call @"Verify_min_balance")
+  mContractRef <- contractCalling @FA2Parameter (Call @"Balance_of") fa2Addr
+  let param = construct (reqs !~ #requests, varExpr selfRef !~ #callback)
+      reqs = (construct (sender, constExpr fa2TokenId)) .: nil
+  ifNone mContractRef (failCustomNoArg @() #noFA2Contract) $
+    transferTokens param (0 mutez)
+  time <- new$ now
+  let startTime = time + (toInt . (#! #unSeconds)) (storage #! #sConfiguration #! #cVoteDelay)
+      endTime = time + (toInt . (#! #unSeconds)) (storage #! #sConfiguration #! #cExpireTime)
```

Comment:
> what is expireTime counted from, the creation time or the voting start time?

from the "creation time".
It's not clarified in every place, but it is unambiguous in the spec:
> After expire_time since the proposal submission, it's no longer possible to vote on it

which, as you noted
> however, that makes it possible to create an unvotable proposal

:thinking:  The current definition seems more intuitive to use to me, although there is the possibility of this mistake.

One option (aside from changing the definition to start from after the `voting_time`), could be to add a check to the `configure` entrypoint, to verify that one is larger than the other.

Reply:
> could be to add a check to the configure entrypoint, to verify that one is larger than the other.

I don't know if it justifies increased gas costs tbh.

Reply:
fair, let's not change this for now then, it's one more thing that's simply a matter of optimization vs convenience really :ok_hand: